File: NeetCode/Stack/Daily_Temperatures/daily__temperatures.py
from typing import Optional, List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [
            0,
        ] * len(temperatures)
        stack = []

        for index, temp in enumerate(temperatures):
            if len(stack) == 0 or stack[-1][1] > temp:
                stack.append((index, temp))
                continue
            while stack and stack[-1][1] < temp:
                idx, tmp = stack.pop()
                res[idx] = index - idx
            stack.append((index, temp))

        return res


if __name__ == "__main__":
    sol = Solution()
    temperatures = [30, 38, 30, 36, 35, 40, 28]
    print(sol.dailyTemperatures(temperatures))
    print("Running Solution...")

File: NeetCode/Stack/Daily_Temperatures/Daily_Temperatures.cpp
// Daily_Temperatures.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> dailyTemperatures(vector<int> &temperatures)
    {
        int n = temperatures.size();
        vector<int> results(n, 0);
        vector<pair<int, int>> stk;
        int tmp, index;
        pair<int, int> tmpPair;
        for (int idx = 0; idx < n; idx++)
        {
            tmp = temperatures[idx];
            if (stk.size() == 0 || stk.back().second > tmp)
            {
                stk.push_back({idx, tmp});
            }
            else
            {
                while (stk.back().second > tmp)
                {
                    tmpPair = stk.back();
                    stk.pop_back();

                    results[tmpPair.first] = idx - tmpPair.first + 1;
                }
            }
            return results;
        }

        return results;
    }
};

int main()
{
    Solution sol;

    vector<int> temperatures = {73, 74, 75, 71, 69, 72, 76, 73};

    vector<int> results = sol.dailyTemperatures(temperatures);

    for (int n : results)
    {
        cout << n << " ";
    }

    return 0;
}
